### PR TITLE
Minimise rendering when text-opacity/opacity is zero

### PR DESCRIPTION
**Pull request by darrenburns**
* For text-opacity we no longer render text as the same colour as the background. We instead render empty cells of the same cell length. This will also improve the "minimum contrast" situation.
* For opacity, the compositor will ignore widgets with 0 opacity (same as when visibility is "hidden" - by still reserving space for the widget).

### CODE REVIEW

File: src/textual/renderables/text_opacity.py [hunk 1]
```diff
@@ -67,11 +68,16 @@ def process_segments(
 
             color = style.color
             bgcolor = style.bgcolor
-            if color and color.triplet and bgcolor and bgcolor.triplet:
-                color_style = _get_blended_style_cached(bgcolor, color, opacity)
-                yield _Segment(text, style + color_style)
+
+            if opacity > 0:
```
Comment (willmcgugan) [inline src/textual/renderables/text_opacity.py:72]: This condition will be constant in the loop. If we move it outside of the loop then we can write an optimized version for the case of opacity == 0

File: src/textual/_compositor.py [hunk 2]
```diff
@@ -604,19 +604,28 @@ def _get_renders(
         # up to this point.
         _rich_traceback_guard = True
 
+        def is_visible(widget: Widget) -> bool:
```
Comment (willmcgugan) [inline src/textual/_compositor.py:607]: This is a fairly tight loop.

Humor me and unroll the code in is_visible. Essentially replace the properties with immediate code, so it makes no implicity function calls.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Looks good, but I'd like to go the extra mile to squeeze a few more cycles out of this.

**Comment by darrenburns** [comment]
@willmcgugan Made those changes, also fixed a link in the docs.

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**